be/src/exec/workgroup/work_group.cpp

```diff
@@ -2,73 +2,241 @@
 
 #include "exec/workgroup/work_group.h"
 
+#include "gen_cpp/internal_service.pb.h"
+#include "glog/logging.h"
 #include "runtime/exec_env.h"
 
 namespace starrocks {
 namespace workgroup {
-WorkGroup::WorkGroup(const std::string& name, int id, size_t cpu_limit, size_t memory_limit, size_t concurrency,
-                     WorkGroupType type)
+
+WorkGroup::WorkGroup(const std::string& name, int64_t id, int64_t version, size_t cpu_limit, double memory_limit,
+                     size_t concurrency, WorkGroupType type)
         : _name(name),
           _id(id),
+          _version(version),
           _cpu_limit(cpu_limit),
           _memory_limit(memory_limit),
           _concurrency(concurrency),
           _type(type) {}
 
+WorkGroup::WorkGroup(const TWorkGroup& twg) : _name(twg.name), _id(twg.id) {
+    if (twg.__isset.cpu_core_limit) {
+        _cpu_limit = twg.cpu_core_limit;
+    } else {
+        _cpu_limit = -1;
+    }
+    if (twg.__isset.mem_limit) {
+        _memory_limit = twg.mem_limit;
+    } else {
+        _memory_limit = -1;
+    }
+    if (twg.__isset.concurrency_limit) {
+        _concurrency = twg.concurrency_limit;
+    } else {
+        _concurrency = -1;
+    }
+    if (twg.__isset.workgroup_type) {
+        _type = twg.workgroup_type;
+    }
+    if (twg.__isset.version) {
+        _version = twg.version;
+    }
+}
+
+TWorkGroup WorkGroup::to_thrift() const {
+    TWorkGroup twg;
+    twg.__set_id(_id);
+    twg.__set_version(_version);
+    return twg;
+}
+
+TWorkGroup WorkGroup::to_thrift_verbose() const {
+    TWorkGroup twg;
+    twg.__set_id(_id);
+    twg.__set_name(_name);
+    twg.__set_version(_version);
+    twg.__set_workgroup_type(_type);
+    std::string state = is_marked_del() ? "dead" : "alive";
+    twg.__set_state(state);
+    twg.__set_cpu_core_limit(_cpu_limit);
+    twg.__set_mem_limit(_memory_limit);
+    twg.__set_concurrency_limit(_concurrency);
+    twg.__set_num_drivers(_acc_num_drivers);
+    return twg;
+}
+
 void WorkGroup::init() {
-    _mem_tracker = std::make_shared<starrocks::MemTracker>(_memory_limit, _name,
-                                                           ExecEnv::GetInstance()->query_pool_mem_tracker());
+    int64_t limit = ExecEnv::GetInstance()->query_pool_mem_tracker()->limit() * _memory_limit;
```

Comment:
I'm confused about the multiplication, could you explain?

Reply:
mem_limit is a ratio